src/query/service/src/storages/stage/stage_table.rs

```diff
@@ -141,75 +140,57 @@ impl Table for StageTable {
 
     fn append2(&self, ctx: Arc<dyn TableContext>, pipeline: &mut Pipeline, _: bool) -> Result<()> {
         let mut sink_pipeline_builder = SinkPipeBuilder::create();
-        for _ in 0..pipeline.output_len() {
-            let input_port = InputPort::create();
-            sink_pipeline_builder.add_sink(
-                input_port.clone(),
-                ContextSink::create(input_port, ctx.clone()),
-            );
+        let single = self.table_info.stage_info.copy_options.single;
+        let op = StageTable::get_op(&ctx, &self.table_info.stage_info)?;
+
+        let uuid = uuid::Uuid::new_v4().to_string();
+        let group_id = AtomicUsize::new(0);
+
+        // parallel compact unload, the partial block will flush into next operator
+        if !single && pipeline.output_len() > 1 {
```

Comment:
Where do we do the parallel compact?

Reply:
if single is not set, it's on parallel.

Reply:
The pipeline only adds one transform with:
```
pipeline.add_transform(|input, output| {
```

So, the parallel is in?

Reply:
If the pipeline's output size is larger than 1, `pipeline.add_transform` will add a transformer for each output  (like TransformLimit).

The idealized parallel can be achieved after #7939 to avoid extra Exchange in cluster queries.